### DISCUSSION TITLE
Adding Nullability Support to DuckDB LogicalType for Better Arrow Schema Compatibility

### DISCUSSION THREAD

**Discussion by rustyconover** [duckdb/duckdb#13949] [Ideas]
In Apache Arrow, each member of a `struct` or `schema` comes with a nullability flag that specifies whether the member can be null. Unfortunately, DuckDB currently doesn’t offer the same capability. Instead, DuckDB enforces column nullability through table constraints.

After spending some time exploring the DuckDB source code, I found that the LogicalType does not include a nullability flag. This means the issue isn't related to the Arrow interface code.

I discovered this limitation while attempting to perform a "round-trip" with an Arrow schema through DuckDB. By "round-trip," I mean creating an Arrow schema, registering it with DuckDB, and then running `select * from [registered table]` to ensure that the result set schema matches the input schema.

Here’s a brief example of how you can manage the nullability flag when creating a Python schema:

```python
import pyarrow as pa
schema = pa.schema([
  pa.field("name", pa.string(), False),
  pa.field("age", pa.string())
])
```

In Arrow, fields are nullable by default, as detailed in the [documentation](https://arrow.apache.org/docs/python/generated/pyarrow.field.html).

If support for nullability were to be introduced in DuckDB, it would likely require updates to the SQL syntax. For instance, creating structs with nullability constraints might look something like this:

```sql
CREATE TABLE data (customer STRUCT(name VARCHAR NOT NULL, age INTEGER));
```

Additional code would also be needed to validate these field-level constraints.   I opened an issue #13947 related to this idea.

**Comment by NickCrews** [comment]
Related issue is https://github.com/duckdb/duckdb/issues/13501 where I was surprised that all `.arrow()` results from duckdb are nullable. But after I looked around the codebase and came to the same conclusion as here, my conclusion was to just give up: getting always-null arrow results just wasn't actually too much of a problem for me. But if someone else wants to implement the suggestion here, that seems nice, but I suspect there are very likely more as-yet-undiscovered problems with that.
